classes/local/rule/rule_base.php

```diff
@@ -51,9 +57,38 @@ abstract class rule_base implements rule_interface {
      */
     public function __construct(stdClass $config) {
         $this->config = $config;
+        $this->id = $this->config->id;
+        $this->type = $this->config->type;
         $this->result = new rule_check_result();
     }
```

Comment:
Yeah on reflection I think first the $config param is badly named, so this needs to be changed at the very least.

Maybe we should be constructing a rule object with the correct properties so it doesn't have to transform them itself internally then?

But something like this maybe doesn't make sense as we may as well just pass it the DB record...

$id, $type, $points, $fallbackpoints, $sortorder, $config

So maybe we should (as per the comment on the `rule_base` constructor, construct the object with the DB record and let it work out what is "config" or not.

Reply:
I have an idea, I will be back shortly with some more code

Reply:
Roughly, we'll construct the rule object with params as follows ($id, $type, $points, $fallbackpoints, $sortorder, $otherconfig). Maybe the constructor can validate `$otherconfig` against what's returned from `get_instance_settings_fields()`.

Reply:
OK so I did this and although there is more code it seems a lot simpler/easier to follow, plus I've thrown a few extra bits of validation in there too. I'll push it in the morning as a new commit.

Reply:
I like the split parameters. What I'm just thinking about is: we currently need the instance configuration, even if there is none. Also our DB column `other` is nullable, but we're assuming everywhere, that there is configuration.

It should at least be optional. Maybe I'd even go as far as to say: the setting of instance configuration should rely upon implementation of the proper interface and a setter-method, called when relevant, instead of the constructor parameter.